New features for the ML-POD package

exapde opened the pull request:
**Summary**

This pull request implements several news features for the ML-POD package for the fitting and calculation of machine-learning interatomic potentials (MLIPs) based on proper orthogonal descriptors (POD). The implementation is located in the folder src/ML-POD and src/KOKKOS, the documentation is located in the folder doc/src, and the examples are provided in examples/PACKAGES/pod. The new features of this pull request include:

- Atom density formulation of many-body proper orthogonal descriptors (POD)
- The environment-adaptive machine learning potentials
- A number of compute classes for computing POD descriptors and their derivatives.
- Per-atom energies and stresses
- Kokkos implementation of the POD potentials
- Group weight functionality allows users to define energy weight and force weight for each exyz input file

**Related Issue(s)**

This pull request does not address any open GitHub issue.

**Author(s)**

Ngoc Cuong Nguyen, Department of Aeronautics and Astronautics, Massachusetts Institute of Technology,
77 Massachusetts Avenue, Cambridge, 02139, MA. Email: [cuongng@mit.edu](mailto:cuongng@mit.edu) and [exapde@gmail.com](mailto:exapde@gmail.com)

Andrew Rohskopf, Sandia National Laboratories. Email: [adrohsk@sandia.gov](mailto:adrohsk@sandia.gov).

**Licensing**

By submitting this pull request, I agree, that my contribution will be included in LAMMPS and redistributed under either the GNU General Public License version 2 (GPL v2) or the GNU Lesser General Public License version 2.1 (LGPL v2.1).

**Backward Compatibility**

This pull request should not break backward compatibility for inputs.

**Implementation Notes**

In this pull request, three C++ classes were created, including eapod, pair_pod, and fitpod_command, to implement the fitting command and pair style for POD. Furthermore, several Compute classes were created to calculate the descriptors and their derivatives. Examples are provided in the examples/Packages/pod folder to demonstrate ML-POD for Ta element and InP compound.

This pull request should not affect other features in LAMMPS.

**Post Submission Checklist**

- [x] The feature or features in this pull request is complete
- [x] Licensing information is complete
- [x] Corresponding author information is complete
- [x] The source code follows the LAMMPS formatting guidelines
- [x] Suitable new documentation files and/or updates to the existing docs are included
- [x] The added/updated documentation is integrated and tested with the documentation build system
- [x] The feature has been verified to work with the conventional build system
- [x] The feature has been verified to work with the CMake based build system
- [x] Suitable tests have been added to the unittest tree.
- [x] A package specific README file has been included or updated
- [x] One or more example input decks are included

**Further Information, Files, and Links**

The original code of this pull request is hosted at https://github.com/cesmix-mit/lammps/tree/kokkospod. The Github repository https://github.com/cesmix-mit/pod_examples contains additional examples for this pull request.

More information about the authors can be found on their personal websites https://www.mit.edu/~cuongng/ and https://rohskopf.github.io/. Technical details about the methodology implemented in this pull request can be found in the papers Journal of Computational Physics, 480, 112030, (2023), Physical Review B, 107(14), 144103, (2023), and https://arxiv.org/abs/2405.00306.

exapde committed:
update pod

exapde committed:
update kokkospod

exapde committed:
clean up redundant variables

exapde committed:
uodate

exapde committed:
fix a bug due to zero neighbor

exapde committed:
fix zero neighbor in compute classes

exapde committed:
fix compute

exapde committed:
fix compute pod

exapde committed:
New force calculation

exapde committed:
Merge branch 'lammps:develop' into kokkospod

exapde committed:
update kokkospod

exapde committed:
Merge branch 'kokkospod' of https://github.com/cesmix-mit/lammps into kokkospod

exapde committed:
documentation and examples

exapde committed:
Delete InP training data

exapde committed:
Merge branch 'lammps:develop' into kokkospod

exapde committed:
Change int* to tagint*

exapde committed:
allocate memory for work

exapde committed:
Initialize sum to zero

exapde committed:
replace malloc with memory->create

exapde committed:
fix whitespace

exapde committed:
Update Ta example

akohlmey reviewed (changes requested):
It would be very helpful if you would change the file names in your examples to follow the example of other example folders. Check out the `flow`, `melt`, `peptide`, or `pour` example folders I.e.
- input files have the name `in.<style>_<info>` where info is a word indicating what the purpose of the job is or something like "first", "second", etc. if jobs need to be run in order
- data files have the name `data.<style>`
- log files have the name `log.<date>.<style>_<info>.<compiler>.<ncpu>`
  typically this would be done with the g++ compiler and 1 and 4 processors (via MPI)
- dump file commands would be commented out
- other outputs would be written to files with ".new" appended so they can be compared to the reference output (without .new).

exapde committed:
Fix input file bug and update examples

exapde commented:
> It would be very helpful if you would change the file names in your examples to follow the example of other example folders. Check out the `flow`, `melt`, `peptide`, or `pour` example folders I.e.
> 
> * input files have the name `in.<style>_<info>` where info is a word indicating what the purpose of the job is or something like "first", "second", etc. if jobs need to be run in order
> * data files have the name `data.<style>`
> * log files have the name `log.<date>.<style>_<info>.<compiler>.<ncpu>`
>   typically this would be done with the g++ compiler and 1 and 4 processors (via MPI)
> * dump file commands would be commented out
> * other outputs would be written to files with ".new" appended so they can be compared to the reference output (without .new).

I changed the file names in the example and fixed the input file bug. However, the code failed at the unit test. The failure seems due to manybody-pair-pod.yaml. I updated the unit test by replacing Ta_coeff.pod and Ta_param.pod in the potentials folder. As a result, the output of the code no longer matches the data in manybody-pair-pod.yaml. How do I update manybody-pair-pod.yaml to fix the issue?  

Thank you very much for your help.

akohlmey commented:
> How do I update manybody-pair-pod.yaml to fix the issue?

When you run the regular test for pair style pod with:
```
ctest -R ManybodyPairStyle:pod -V
```
Then you can update the yaml file with:
```
TEST_ARGS=-u ctest -R ManybodyPairStyle:pod -V
```
Then you need to do a couple of manual edits:
- you need to delete the line "tags: generated" (this is a marker for newly updated yaml files)
- you need to change "skip_tests:" to "skip_tests: kokkos_omp" since the Kokkos support does not include multi-threading

I am attaching a patch file that has these changes: 
[update-pod-unit-test.patch.gz](https://github.com/lammps/lammps/files/15410742/update-pod-unit-test.patch.gz)

There are multiple concerns.
- the values for energy, stress, and forces are quite different. That suggests that there is a problem with either the old implementation or the new implementation or that the training data was not suitable for the geometry of the test and thus the potential is not suitable for the test.
- the new implementation fails on the stress output with `pair_modify nofdotr`. The previous implementation passed.

exapde committed:
fix fdotr and update the force-styles unittest

exapde commented:
> When you run the regular test for pair style pod with:
> 
> ```
> ctest -R ManybodyPairStyle:pod -V
> ```
> 
> Then you can update the yaml file with:
> 
> ```
> TEST_ARGS=-u ctest -R ManybodyPairStyle:pod -V
> ```
> 
> Then you need to do a couple of manual edits:
> 
> * you need to delete the line "tags: generated" (this is a marker for newly updated yaml files)
> * you need to change "skip_tests:" to "skip_tests: kokkos_omp" since the Kokkos support does not include multi-threading
> 
> I am attaching a patch file that has these changes: [update-pod-unit-test.patch.gz](https://github.com/lammps/lammps/files/15410742/update-pod-unit-test.patch.gz)
> 
> There are multiple concerns.
> 
> * the values for energy, stress, and forces are quite different. That suggests that there is a problem with either the old implementation or the new implementation or that the training data was not suitable for the geometry of the test and thus the potential is not suitable for the test.
> * the new implementation fails on the stress output with `pair_modify nofdotr`. The previous implementation passed.

Thanks so much Axel. With your help, I was able to fix the fdor issue and update the unit test. All checks passed! I also changed the potential parameters to make the values for energy, stress, and forces reasonable for the geometry of the test.

exapde committed:
update compute descriptors

akohlmey committed:
avoid namespace pollutions from defines in eapot.h

akohlmey committed:
initialize all pointers to null, reorder to match definition

akohlmey committed:
add braces to group commands according to indentation

akohlmey committed:
use utils::numeric to convert text to numbers

akohlmey committed:
small programming style updates

akohlmey committed:
Merge branch 'develop' into kokkospod

akohlmey commented:
@exapde When I enable all warnings (-Wall -Wextra -Wno-maybe-uninitialized) I see a lot of warnings about unused parameters. Wherever possible and meaningful, those parameters should be remove from the list of arguments of those methods. They are just dead weight. 

```
/home/akohlmey/compile/lammps/src/ML-POD/eapod.cpp: In member function ‘void LAMMPS_NS::EAPOD::peratombase_descriptors(double*, double*, double*, double*, int*, int*, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/eapod.cpp:851:14: warning: unused parameter ‘ti’ [-Wunused-parameter]
  851 |         int *ti, int *tj, int Nj)
      |         ~~~~~^~
/home/akohlmey/compile/lammps/src/ML-POD/eapod.cpp: In member function ‘void LAMMPS_NS::EAPOD::radialangularbasis(double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, double*, int*, int, int, int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/eapod.cpp:2249:26: warning: unused parameter ‘tm’ [-Wunused-parameter]
 2249 |                  double *tm, int *atomtype, int N, int K, int M, int Ne)
      |                  ~~~~~~~~^~
/home/akohlmey/compile/lammps/src/ML-POD/fitpod_command.cpp: In member function ‘void LAMMPS_NS::FitPOD::get_exyz_files(std::vector<std::__cxx11::basic_string<char> >&, std::vector<std::__cxx11::basic_string<char> >&, const std::string&, const std::string&)’:
/home/akohlmey/compile/lammps/src/ML-POD/fitpod_command.cpp:377:11: warning: unused variable ‘ext_size’ [-Wunused-variable]
  377 |       int ext_size = extension.size() + 1;
      |           ^~~~~~~~
/home/akohlmey/compile/lammps/src/ML-POD/fitpod_command.cpp: In member function ‘void LAMMPS_NS::FitPOD::energyforce_calculation(const datastruct&, double*)’:
/home/akohlmey/compile/lammps/src/ML-POD/fitpod_command.cpp:1882:70: warning: unused parameter ‘coeff’ [-Wunused-parameter]
 1882 | void FitPOD::energyforce_calculation(const datastruct &data, double *coeff)
      |                                                              ~~~~~~~~^~~~~
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp: In member function ‘void LAMMPS_NS::PairPOD::NeighborCount(double**, int**, int*, int*, double, int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp:384:114: warning: unused parameter ‘gi2’ [-Wunused-parameter]
  384 | void PairPOD::NeighborCount(double **x, int **firstneigh, int *ilist, int *numneigh, double rcutsq, int gi1, int gi2)
      |                                                                                                              ~~~~^~~
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp: In member function ‘void LAMMPS_NS::PairPOD::NeighborList(double**, int**, int*, int*, int*, int*, double, int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp:416:87: warning: unused parameter ‘gi2’ [-Wunused-parameter]
  416 |                                int *ilist, int *numneigh, double rcutsq, int gi1, int gi2)
      |                                                                                   ~~~~^~~
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp: In member function ‘void LAMMPS_NS::PairPOD::radialangularsum2(int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp:962:45: warning: unused parameter ‘Nij’ [-Wunused-parameter]
  962 | void PairPOD::radialangularsum2(int Ni, int Nij)
      |                                         ~~~~^~~
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp: In member function ‘void LAMMPS_NS::PairPOD::twobodydescderiv(double*, int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp:1002:49: warning: unused parameter ‘Ni’ [-Wunused-parameter]
 1002 | void PairPOD::twobodydescderiv(double *dd2, int Ni, int Nij)
      |                                             ~~~~^~
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp: In member function ‘void LAMMPS_NS::PairPOD::threebodydescderiv(double*, int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp:1076:51: warning: unused parameter ‘Ni’ [-Wunused-parameter]
 1076 | void PairPOD::threebodydescderiv(double *dd3, int Ni, int Nij)
      |                                               ~~~~^~
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp: In member function ‘void LAMMPS_NS::PairPOD::fourbodydescderiv(double*, int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp:1325:50: warning: unused parameter ‘Ni’ [-Wunused-parameter]
 1325 | void PairPOD::fourbodydescderiv(double *dd4, int Ni, int Nij)
      |                                              ~~~~^~
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp: In member function ‘void LAMMPS_NS::PairPOD::blockatom_forces(double*, int, int)’:
/home/akohlmey/compile/lammps/src/ML-POD/pair_pod.cpp:2035:11: warning: unused variable ‘cb2’ [-Wunused-variable]
 2035 |   double *cb2 =  &cb[0]; // nl3
      |           ^~~
[ 22%] Building CXX object CMakeFiles/lammps.dir/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp.o
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp: In instantiation of ‘void LAMMPS_NS::PairPODKokkos<DeviceType>::radialbasis(t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, double, double, int, int, int, int, int) [with DeviceType = Kokkos::OpenMP; t_pod_1d = Kokkos::View<double*, Kokkos::OpenMP>]’:
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp:1887:16:   required from here
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp:648:49: warning: unused parameter ‘l_ns’ [-Wunused-parameter]
  648 |     int l_inversedegree, int l_nbesselpars, int l_ns,  int Nij)
      |                                             ~~~~^~~~
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp: In instantiation of ‘void LAMMPS_NS::PairPODKokkos<DeviceType>::fourbody_forces(t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1i, t_pod_1i, t_pod_1i, t_pod_1i, t_pod_1i, t_pod_1i, int, int, int, int, int, int, int, int) [with DeviceType = Kokkos::OpenMP; t_pod_1d = Kokkos::View<double*, Kokkos::OpenMP>; t_pod_1i = Kokkos::View<int*, Kokkos::OpenMP>]’:
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp:1887:16:   required from here
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp:1115:111: warning: unused parameter ‘l_elemindex’ [-Wunused-parameter]
 1115 |     t_pod_1d l_sumU, t_pod_1i l_idxi, t_pod_1i l_tj, t_pod_1i l_pa4, t_pod_1i l_pb4, t_pod_1i l_pc4, t_pod_1i l_elemindex,
      |                                                                                                      ~~~~~~~~~^~~~~~~~~~~
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp: In instantiation of ‘void LAMMPS_NS::PairPODKokkos<DeviceType>::allbody_forces(t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1d, t_pod_1i, t_pod_1i, int, int, int, int, int) [with DeviceType = Kokkos::OpenMP; t_pod_1d = Kokkos::View<double*, Kokkos::OpenMP>; t_pod_1i = Kokkos::View<int*, Kokkos::OpenMP>]’:
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp:1887:16:   required from here
/home/akohlmey/compile/lammps/src/KOKKOS/pair_pod_kokkos.cpp:1336:71: warning: unused parameter ‘l_nabf3’ [-Wunused-parameter]
 1336 |     t_pod_1i l_idxi, t_pod_1i l_tj, int l_nelements, int l_nrbf3, int l_nabf3, int l_K3, int Nij)
```

akohlmey commented:
Before doing a final review (programming style, clang-format, documentation), @stanmoore1 has to look over the KOKKOS part.

akohlmey commented on src/ML-POD/compute_pod_local.h:
```diff
@@ -0,0 +1,60 @@
+
+#ifdef COMPUTE_CLASS
+// clang-format off
+ComputeStyle(pod/ldd,ComputePODLocal);
```
Either the command should be called compute pod/local or the class name should be changed to ComputePODLDD and the files to compute_pod_ldd.h/cpp
If the latter is done, also documentation and possibly examples need to be updated.

exapde replied:
The command is changed to compute pod/local. I updated the documentation and example to be consistent with the new command name.

akohlmey commented on src/ML-POD/compute_pod_global.h:
```diff
@@ -0,0 +1,60 @@
+
+#ifdef COMPUTE_CLASS
+// clang-format off
+ComputeStyle(pod/gdd,ComputePODGlobal);
```
Same as for compute pod/ldd, either the command name should be changed or class and file.

exapde replied:
The command is changed to compute pod/global. I updated the documentation and example to be consistent with the new command name.

akohlmey committed:
spelling updates

akohlmey reviewed (changes requested):
style name, file name, class name and include guard should be consistent

stanmoore1 committed:
Comment out unused var

stanmoore1 committed:
Update .gitignore for POD

stanmoore1 committed:
Add logic for GNU Make

stanmoore1 commented:
@exapde this gives correct numerics on an V100 GPU along with a significant speedup, nice job! I submitted a pull request with a few changes to your fork here: https://github.com/cesmix-mit/lammps/pull/15 since I can't push directly.

stanmoore1 committed:
Small cleanup

stanmoore1 committed:
Update docs

stanmoore1 commented on src/KOKKOS/pair_pod_kokkos.cpp:
```diff
@@ -0,0 +1,1890 @@
+template<class DeviceType>
+PairPODKokkos<DeviceType>::~PairPODKokkos()
+{
+  if (timing == 1) {
```
Is this debugging output that needs to be removed? If not, only rank 0 should print to the screen, e.g. an average time across all ranks.

exapde replied:
I turn off the timing and make rank 0 print to the screen.

stanmoore1 commented on src/KOKKOS/pair_pod_kokkos.cpp:
```diff
@@ -0,0 +1,1890 @@
+template<class DeviceType>
+void PairPODKokkos<DeviceType>::compute(int eflag_in, int vflag_in)
+{
+//   if (host_flag) {
```
Please remove commented out code if it is not needed.

exapde replied:
The commented out code is removed.

stanmoore1 reviewed (commented):
Kokkos part looks good, I had a few minor change requests.

exapde committed:
Merge pull request #15 from stanmoore1/kokkospod

Small changes for https://github.com/lammps/lammps/pull/4168

exapde committed:
Remove unused parameters, change command name, and update documentation and examples..

exapde committed:
Merge branch 'kokkospod' of https://github.com/cesmix-mit/lammps into kokkospod

exapde committed:
Remove commented out code and turn off timing

exapde commented:
Hi Axel and Stan!

Thank you very much for reviewing the code and suggesting changes for improvement. I believe that I have made all the changes you suggested. Please let me know if you have any other suggestions.

Cheers, Cuong.

akohlmey committed:
Merge commit 'refs/pull/4168/head' of github.com:lammps/lammps into kokkospod

akohlmey commented:
> I believe that I have made all the changes you suggested. Please let me know if you have any other suggestions.

You've missed merging *my* pull request (#14) in your repository. It had some merge conflict after addressing Stan's requests for changes, but I have resolved them now and it is mergable again.

exapde committed:
Merge pull request #14 from akohlmey/kokkospod

Update kokkospod branch to upstream and minor cleanups and corrections

exapde commented:
> You've missed merging _my_ pull request (#14) in your repository. It had some merge conflict after addressing Stan's requests for changes, but I have resolved them now and it is mergable again.

Sorry I missed that. I just merge your pull request.

akohlmey commented:
> Please let me know if you have any other suggestions.

You are still using printf() in many places:
```
$ grep printf ML-POD/*.cpp
ML-POD/eapod.cpp:        printf("reallocate temporary memory with Njmax = %d ...\n", Njmax);
ML-POD/eapod.cpp:        printf("reallocate temporary memory with Njmax = %d ...\n", Njmax);
ML-POD/eapod.cpp:        printf("reallocate temporary memory with Njmax = %d ...\n", Njmax);
ML-POD/eapod.cpp:    printf( "\n %s\n", desc );
ML-POD/eapod.cpp:        for( j = 0; j < n; j++ ) printf( " %d", a[i+j*lda] );
ML-POD/eapod.cpp:        printf( "\n" );
ML-POD/eapod.cpp:    printf( "\n %s\n", desc );
ML-POD/eapod.cpp:        for( j = 0; j < n; j++ ) printf( " %6.12f", a[i+j*lda] );
ML-POD/eapod.cpp:        printf( "\n" );
ML-POD/fitpod_command.cpp:  //printf("number of configurations = %d\n", (int) data.num_atom.size());
ML-POD/fitpod_command.cpp://   printf("%d %d %d\n",comm->me, nElemAtomsCount[0], nElemAtomsCount[1]);
ML-POD/fitpod_command.cpp:    //for (int i = 0; i < desc.nClusters * nComponents; i++) printf("centroids[%d] = %f\n", i, centroids[i]);
ML-POD/fitpod_command.cpp:  printf( "\n %s\n", desc );
ML-POD/fitpod_command.cpp:    for( j = 0; j < n; j++ ) printf( " %6.12f", a[j][i] );
ML-POD/fitpod_command.cpp:    printf( "\n" );
ML-POD/fitpod_command.cpp:  printf( "\n %s\n", desc );
ML-POD/fitpod_command.cpp:    for( j = 0; j < n; j++ ) printf( " %6.12f", a[i+j*lda] );
ML-POD/fitpod_command.cpp:    printf( "\n" );
ML-POD/fitpod_command.cpp:  printf( "\n %s\n", desc );
ML-POD/fitpod_command.cpp:    for( j = 0; j < n; j++ ) printf( " %d", a[i+j*lda] );
ML-POD/fitpod_command.cpp:    printf( "\n" );
ML-POD/pair_pod.cpp:    printf("Interval size must be a positive integer.\n");
```
This should not be done since you cannot turn this off. Instead, you should be using `fprintf(screen,...)` and `fprintf(logfile,...)` or - for convenience - `utilslogmesg()` this will output to both and supports `fmt::format` formatting.
Output for all of these is usually only done by MPI rank zero to avoid having many lines of output for large parallel jobs (very large ones may us use 100s of thousands of MPI ranks).

Please see: https://docs.lammps.org/Modify_style.html#miscellaneous-standards-varied
and see: https://docs.lammps.org/Developer_code_design.html#i-o-and-output-formatting

P.S.: there still seems to be commended out code that should be removed.

exapde committed:
remove printf or replace it with utils:logmsg

exapde committed:
Merge branch 'kokkospod' of https://github.com/cesmix-mit/lammps into kokkospod

exapde committed:
Remove commented out code

exapde commented:
Hi Axel!

Thanks very much for your review. I removed the commented our code and replaced fprintf. 

C

akohlmey commented:
> Thanks very much for your review. I removed the commented our code and replaced fprintf.

@exapde I am sorry that my comments are trickling in in an annoyingly piece-wise fashion (a bit like the proverbial Chinese water torture), but while I am checking out your recent changes, I am running across items that I didn't notice before. What I am looking at now are rather cosmetic and non-functional changes. Unfortunately, your GitHub account doesn't allow me push such commits directly into your pull request. So here is a question for you: do you want me to list all the details on the pull request or should we merge the pull request as is and then my changes will wind up in PR #4203?

Please let us know.

@stanmoore1 you can start your final review now.

exapde commented:
> Please let us know.

Let's merge the pull request as is.

exapde commented:
Thanks @akohlmey and @stanmoore1 for the careful review that help me write better code.

stanmoore1 commented on src/ML-POD/eapod.cpp:
```diff
@@ -0,0 +1,4734 @@
+  init3body(P3);
+  init4body(P4);
+
+  int nb[] = {1,     2,     4,     7,    11,    16,    23};
```
This formatting is a little odd, but maybe Axel is going to run this through clang format?

akohlmey replied:
yes, among other things
